docs/DifferentiableTypes.md

```diff
@@ -0,0 +1,289 @@
+    }
+}
+
+// Differential operators like `gradient(at:in:)` just work!
+let dense = DenseLayer(weight: [[1, 1], [1, 1]], bias: [0, 0])
+let 𝛁dense = gradient(at: dense) { dense in dense.applied(to: [[3, 3]]).sum() }
+
+dump(𝛁dense)
+// ▿ DenseLayer.AllDifferentiableVariables
+//   - weight: [[3.0, 3.0], [3.0, 3.0]]
+//   - bias: [1.0, 1.0]
+```
+
+## Protocol details
+
+Here is the full `Differentiable` protocol definition. More explanation is provided below.
+
+```swift
+/// A type that mathematically represents a differentiable manifold whose
+/// tangent spaces are finite-dimensional.
+public protocol Differentiable {
+    /// The tangent bundle of this differentiable manifold.
+    associatedtype TangentVector : AdditiveArithmetic & Differentiable
+        where TangentVector.TangentVector == TangentVector,
+              TangentVector.CotangentVector == CotangentVector,
+              TangentVector.AllDifferentiableVariables == TangentVector
+    /// The cotangent bundle of this differentiable manifold.
+    associatedtype CotangentVector : AdditiveArithmetic & Differentiable
+        where CotangentVector.TangentVector == CotangentVector,
+              CotangentVector.CotangentVector == TangentVector,
+              CotangentVector.AllDifferentiableVariables == CotangentVector
+    /// The type of all differentiable variables in this type.
+    associatedtype AllDifferentiableVariables : Differentiable
+        where AllDifferentiableVariables.AllDifferentiableVariables ==
+                  AllDifferentiableVariables,
+              AllDifferentiableVariables.TangentVector == TangentVector,
+              AllDifferentiableVariables.CotangentVector == CotangentVector
+
+    /// All differentiable variables in this type.
+    var allDifferentiableVariables: AllDifferentiableVariables { get }
+
+    /// Returns `self` moved along the value space towards the given tangent
+    /// vector. In Riemannian geometry (mathematics), this represents an
+    /// exponential map.
+    func moved(along direction: TangentVector) -> Self
+
+    /// Convert a cotangent vector to its corresponding tangent vector.
+    func tangentVector(from cotangent: CotangentVector) -> TangentVector
+}
+```
+
+Mathematically, `Differentiable` represents a differentiable manifold: this is a technical term for smooth-surfaced objects like spheres and generalizes types that are compatible with differentiation, like `Float`, `Double`, `Tensor`, and `SIMD4<Float>`. This definition comes from differential geometry and is quite technical, and not all details are relevant for most use cases.
+
+<p align="center">
+  <img src="https://upload.wikimedia.org/wikipedia/commons/3/37/Pushforward.svg" align=center>
+  <br>
+  <sub>Image showing two differentiable manifolds: a sphere and a spheroid.</sub>
+  <br>
+  <sub>From https://en.wikipedia.org/wiki/Pushforward_(differential).</sub>
+</p>
+
+Here are the important points about the `Differentiable` protocol:
```

Comment:
Hmm... "Points" sounds pretty weird to me since you are iterating over every requirement and explaining them.